src/libraries/System.Net.Http/src/System/Net/Http/HttpRequestMessage.cs

```diff
@@ -197,7 +200,40 @@ private void InitializeValues(HttpMethod method, Uri? requestUri)
 
         internal bool MarkAsSent()
         {
-            return Interlocked.Exchange(ref _sendStatus, MessageAlreadySent) == MessageNotYetSent;
+            return Interlocked.CompareExchange(ref _sendStatus, MessageAlreadySent, MessageNotYetSent) == MessageNotYetSent;
+        }
+
+        internal void MarkAsTrackedByTelemetry()
+        {
+            if (_finalizer is null)
+            {
+                _finalizer = new HttpRequestMessageFinalizer();
+            }
+            else
+            {
+                GC.ReRegisterForFinalize(_finalizer);
+            }
+
+            Debug.Assert(_sendStatus != MessageShouldEmitTelemetry);
+            _sendStatus = MessageShouldEmitTelemetry;
```

Comment:
Currently assumes that an `HttpRequestMessage` instance will not be used for multiple requests in parallel. We could guard against that by using Interlocked here.

Reply:
If this is the case, does `MarkAsSent` need to use interlocked?

Reply:
It does if we want to keep the guarantee of the request only being used once by HttpClient. As this restriction is only imposed by HttpClient, other HttpMessageInvokers are free to reuse the request right now, just not in parallel.
We could choose to provide the same level of thread-safety for both scenarios.